### PR TITLE
Timezone and fraction seconds fixes and tests

### PR DESCRIPTION
**Pull request by grajoria**
Fix for fraction seconds accuracy.
Added Support for nano seconds.
Initializing timezone between rows while running multi-row query.
Added checks for negative timezone.
Added nested queries tests for single and multi row queries.

### CODE REVIEW

File: include/s3select_functions.h [hunk 1]
```diff
@@ -25,65 +25,91 @@ struct push_char
...
 
 };
-static push_2dig g_push_2dig;
+static push_4dig g_push_4dig;
 
-struct push_3dig
+struct push_1fdig
 {
-  void operator()(const char* a, const char* b, uint32_t* n) const
+  void operator()(const char* a, const char* b, double* n) const
```
Comment (galsalomon66) [inline include/s3select_functions.h:50]: no need for floating-point, it's less accurate.
the nano-sec should be uint32 (amount of nano-sec, not a fraction)
the push_Xfdig::operator() should multiply the result by 10^(6- number of digits) i.e. 1,10,100,...

File: include/s3select_functions.h [hunk 2]
```diff
@@ -781,9 +811,10 @@ struct _fn_to_timestamp : public base_function
     hr = 0;
     mn = 0;
     sc = 0;
-    frac_sec = 0;
+    frac_sec = 0.0;
```
Comment (galsalomon66) [inline include/s3select_functions.h:814]: uint32 instead of double

File: include/s3select_functions.h [hunk 3]
```diff
@@ -822,24 +855,16 @@ struct _fn_to_timestamp : public base_function
                           boost::posix_time::hours(hr) +
                           boost::posix_time::minutes(mn) +
                           boost::posix_time::seconds(sc) +
-                          boost::posix_time::nanoseconds(frac_sec*1000000000));
+                          boost::posix_time::nanoseconds(frac_sec * 1000000000));
```
Comment (galsalomon66) [inline include/s3select_functions.h:858]: once using uint, not need for multiplication.

**Comment by grajoria** [comment]
@galsalomon66 Please review the updated patch

**Comment by galsalomon66** [comment]
the PR should take care also of build-option for nano/micro 
https://www.boost.org/doc/libs/1_53_0/doc/html/date_time/details.html#date_time.buildinfo
(use of BOOST_DATE_TIME_POSIX_TIME_STD_CONFIG macro)

**Comment by grajoria** [comment]
@galsalomon66 Added nano seconds build option and updated tested also. Please review

File: CMakeLists.txt [hunk 4]
```diff
@@ -9,6 +9,8 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
 find_package(Boost REQUIRED)
 find_package(GTest REQUIRED)
 
+add_definitions(-DBOOST_DATE_TIME_POSIX_TIME_STD_CONFIG)
```
Comment (galsalomon66) [inline CMakeLists.txt:12]: the nanosec should added conditionally.

if(DEFINED ENV{XXX})
  add_definitions(-DBOOST_DATE_TIME_POSIX_TIME_STD_CONFIG)
endif()

Comment (galsalomon66) [reply CMakeLists.txt:12]: (this works only for standalone application, for CEPH, it need another definition, will not do that)

File: include/s3select_functions.h [hunk 5]
```diff
@@ -15,6 +15,11 @@ using namespace std::string_literals;
 namespace s3selectEngine
 {
 
+constexpr double sec_scale(int n)
+{
+    return pow(10, n);
+}
+
```
Comment (galsalomon66) [inline include/s3select_functions.h:22]: #if BOOST_DATE_TIME_POSIX_TIME_STD_CONFIG
      const double scale = sec_scale(9-1);    //nano-sec
    #else
      const double scale = sec_scale(6-1);    //micro-sec
    #endif
    
    no need to repeat 9 times this part, 
    /const double scale/ could be defined globally(and static)

File: include/s3select_functions.h [hunk 6]
```diff
@@ -673,28 +782,37 @@ struct _fn_to_timestamp : public base_function
...
   bsc::rule<> d_yyyy_dig = ((dig4[BOOST_BIND_ACTION_PARAM(push_4dig, &yr)]) >> *(delimiter));
   bsc::rule<> d_yyyymmdd_dig = ((dig4[BOOST_BIND_ACTION_PARAM(push_4dig, &yr)]) >> *(date_separator)
                                 >> (dig2[BOOST_BIND_ACTION_PARAM(push_2dig, &mo)]) >> *(date_separator)
                                 >> (dig2[BOOST_BIND_ACTION_PARAM(push_2dig, &dy)]) >> *(delimiter));
 
-  uint32_t hr = 0, mn = 0, sc = 0, frac_sec = 0, tz_hr = 0, tz_mn = 0, sign, tm_zone;
+  uint32_t hr = 0, mn = 0, sc = 0,  frac_sec = 0, tz_hr = 0, tz_mn = 0, sign, tm_zone = '0';
+  bsc::rule<> fdig9 = bsc::lexeme_d[bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p];
+  bsc::rule<> fdig8 = bsc::lexeme_d[bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p];
+  bsc::rule<> fdig7 = bsc::lexeme_d[bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p >> bsc::digit_p];
```
Comment (galsalomon66) [inline include/s3select_functions.h:796]: the /fdig7-9/ rules should be under compile flag.

File: include/s3select_functions.h [hunk 7]
```diff
@@ -673,28 +782,37 @@ struct _fn_to_timestamp : public base_function
...
   bsc::rule<> d_timezone_dig =  ((timezone_sign[BOOST_BIND_ACTION_PARAM(push_char, &sign)]) >> (dig2[BOOST_BIND_ACTION_PARAM(push_2dig, &tz_hr)]) >> *(time_separator)
                                 >> (dig2[BOOST_BIND_ACTION_PARAM(push_2dig, &tz_mn)])) | (zero_timezone[BOOST_BIND_ACTION_PARAM(push_char, &tm_zone)]);
 
-  bsc::rule<> fraction_sec = (dig6[BOOST_BIND_ACTION_PARAM(push_6dig, &frac_sec)]) |
-                             (dig5[BOOST_BIND_ACTION_PARAM(push_5dig, &frac_sec)]) |
-			     (dig4[BOOST_BIND_ACTION_PARAM(push_4dig, &frac_sec)]) |
-			     (dig3[BOOST_BIND_ACTION_PARAM(push_3dig, &frac_sec)]) |
-			     (dig2[BOOST_BIND_ACTION_PARAM(push_2dig, &frac_sec)]) |
-			     (dig1[BOOST_BIND_ACTION_PARAM(push_1dig, &frac_sec)]);
+  bsc::rule<> fraction_sec = (fdig9[BOOST_BIND_ACTION_PARAM(push_9fdig, &frac_sec)]) |
```
Comment (galsalomon66) [inline include/s3select_functions.h:807]: fraction_sec should be defined with/without  /fdig7 .. 9/ rules, according to compile-flag.
the fractionn_sec is less complex in that way.

currently, the grammar  accepts nano-sec(9 digits), while it should do that only upon nano-sec setup.

File: include/s3select_functions.h [hunk 8]
```diff
@@ -15,6 +15,11 @@ using namespace std::string_literals;
 namespace s3selectEngine
 {
 
+constexpr double sec_scale(int n)
```
Comment (galsalomon66) [inline include/s3select_functions.h:18]: :+1:

File: include/s3select_functions.h [hunk 9]
```diff
@@ -817,29 +938,21 @@ struct _fn_to_timestamp : public base_function
...
     boost::posix_time::ptime new_ptime;
 
-    #if NANO_SEC
+    #if BOOST_DATE_TIME_POSIX_TIME_STD_CONFIG
       new_ptime = boost::posix_time::ptime(boost::gregorian::date(yr, mo, dy),
                           boost::posix_time::hours(hr) +
                           boost::posix_time::minutes(mn) +
                           boost::posix_time::seconds(sc) +
-                          boost::posix_time::nanoseconds(frac_sec*1000000000));
+                          boost::posix_time::nanoseconds(frac_sec));
```
Comment (galsalomon66) [inline include/s3select_functions.h:946]: :+1:

File: test/s3select_test.cpp [hunk 10]
```diff
@@ -348,7 +349,11 @@ TEST(TestS3selectFunctions, to_timestamp)
     EXPECT_EQ(s3select_res, out_timestamp);
 
     timestamp = "2007-09-17T17:56:05.234Z";
-    out_timestamp = "2007-09-17T17:56:05.234000Z";
+    #if BOOST_DATE_TIME_POSIX_TIME_STD_CONFIG
```
Comment (galsalomon66) [inline test/s3select_test.cpp:352]: :+1:

File: CMakeLists.txt (hunk 4 above)
Comment (grajoria) [reply CMakeLists.txt]: Done

File: include/s3select_functions.h (hunk 6 above)
Comment (grajoria) [reply include/s3select_functions.h]: Done

File: include/s3select_functions.h (hunk 7 above)
Comment (grajoria) [reply include/s3select_functions.h]: Done

File: CMakeLists.txt [hunk 11]
```diff
@@ -9,6 +9,10 @@ set(CMAKE_CXX_STANDARD_REQUIRED ON)
 find_package(Boost REQUIRED)
 find_package(GTest REQUIRED)
 
+if(DEFINED ENV{NANO_SEC})
```
Comment (galsalomon66) [inline CMakeLists.txt:12]: :+1:

File: include/s3select_functions.h [hunk 12]
```diff
@@ -25,65 +30,156 @@ struct push_char
...
-static push_2dig g_push_2dig;
+static push_4dig g_push_4dig;
 
-struct push_3dig
+struct push_1fdig
 {
   void operator()(const char* a, const char* b, uint32_t* n) const
   {
-    *n = (static_cast<char>(*a) - 48) * 100 + (static_cast<char>(*(a+1)) - 48) * 10 + (static_cast<char>(*(a+2)) - 48);
+    #if BOOST_DATE_TIME_POSIX_TIME_STD_CONFIG
```
Comment (galsalomon66) [inline include/s3select_functions.h:57]: :+1:

**Review by galsalomon66 — COMMENTED** [review]
please check the following queries results
select to_string(to_timestamp("2007-09-17T17:56:05.003456+00:00"),"n") from

select to_string(to_timestamp("2007-09-17T17:56:05.3456+00:00"),"n") from

**Comment by grajoria** [comment]
> please check the following queries results
> select to_string(to_timestamp("2007-09-17T17:56:05.003456+00:00"),"n") from
> 
> select to_string(to_timestamp("2007-09-17T17:56:05.3456+00:00"),"n") from

I tried queries like you mentioned above. Both the queries were giving same results, which should not happen.
I have made changes for it in to_string function for formats 'n' and 'S...S' and added tests to cover this case. The PR is updated with the same.

**Pull request merged**